nomos-services/cryptarchia-consensus/src/leadership.rs

```diff
@@ -1,59 +1,112 @@
-    pub fn follow_chain(&mut self, parent_id: HeaderId, id: HeaderId, to_evolve: Commitment) {
-        if let Some(coins) = self.coins.get(&parent_id) {
-            let coins = coins
+    pub fn follow_chain(&mut self, parent_id: HeaderId, id: HeaderId, to_evolve: Nullifier) {
+        if let Some(notes) = self.notes.get(&parent_id) {
+            let notes = notes
                 .iter()
-                .map(|coin| {
-                    if coin.commitment() == to_evolve {
-                        coin.evolve()
+                .map(|note| {
+                    if note.nullifier() == to_evolve {
+                        evolve(note)
                     } else {
-                        *coin
+                        *note
                     }
                 })
                 .collect();
-            self.coins.insert(id, coins);
+            self.notes.insert(id, notes);
         }
     }
 
-    pub fn build_proof_for(
+    pub async fn build_proof_for(
         &self,
+        note_tree: &NoteTree,
         epoch_state: &EpochState,
         slot: Slot,
         parent: HeaderId,
-    ) -> Option<LeaderProof> {
-        let coins = self.coins.get(&parent)?;
-        for coin in coins {
-            if coin.is_slot_leader(epoch_state, slot, &self.config.consensus_config) {
+    ) -> Option<Risc0LeaderProof> {
+        let notes = self.notes.get(&parent)?;
+        for note in notes {
+            let Some(index) = note_tree
+                .commitments()
+                .iter()
+                .position(|cm| cm == &note.note_commitment())
+            else {
+                continue;
+            };
+
+            let input_cm_path = note_tree
+                .witness(index)
+                .expect("Note was found in the tree");
+            let public_inputs = LeaderPublic::new(
+                note_tree.root(),
+                *epoch_state.nonce(),
+                slot.into(),
+                self.config.consensus_config.active_slot_coeff,
+                epoch_state.total_stake(),
+                note.nullifier(),
+                note.evolve_output(b"NOMOS_POL").commit_note(),
+            );
+            if public_inputs.check_winning(note) {
                 tracing::debug!(
                     "leader for slot {:?}, {:?}/{:?}",
                     slot,
-                    coin.value(),
+                    note.note.value,
                     epoch_state.total_stake()
                 );
-                let proof = coin.to_proof(slot);
-                return Some(proof);
+                let input = *note;
+                let res = tokio::task::spawn_blocking(move || {
+                    Risc0LeaderProof::build(
+                        public_inputs,
+                        LeaderPrivate {
+                            input,
+                            input_cm_path,
+                        },
+                    )
+                })
+                .await;
```

Comment:
Is this a heavy operation? I don't get the spawn_blocking to immediately await.

Reply:
yes, this is heavy and cpu intenstive (i.e. does not yield).`spawn_blocking` uses a dedicated thread for this to avoid blocking other tasks